### PR TITLE
nexty governance smart contract and dapp

### PR DESCRIPTION
**Pull request by hadv**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,158 +1,9 @@
...
-
-```
-npx truffle console --network ganache
-truffle(ganache)> token = NTFToken.at("0x6e27d4f5a5216388fe555c59dc20b5d1c1d12406");
-truffle(ganache)> token.getHolder(0);
-truffle(ganache)> token.getHolder(1);
-```
-
-Congratulation! Finally, we can work with new smart contract logic and can still preserve the state from old smart contract.
+Mainet Nexty
```
Comment (hadv) [inline README.md]: why did you remove my readme? please get back!

File: contracts/governance/NextyManager.sol [hunk 2]
```diff
@@ -0,0 +1,265 @@
+//solium-disable linebreak-style
+pragma solidity ^0.4.24;
+
+import "zos-lib/contracts/migrations/Migratable.sol";
+import "openzeppelin-zos/contracts/ownership/Ownable.sol";
+import "./SafeMath.sol";
+import "./Blacklist.sol";
+
+contract NTFToken {
```
Comment (hadv) [inline contracts/governance/NextyManager.sol]: remove unused contract; you can change type of `ntfAddress` to ERC20

File: contracts/governance/NextyManager.sol [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        //ntf amount deposited
+        uint256 balance;
+        //sealer used address to seal blocks
+        address coinbase;
+        //withdrawable time after leaving
+        uint256 unlockTime;
+    }
+
+    //ntf token contract, unit used to join Nexty sealers
+    address public ntfAddress;
```
Comment (hadv) [inline contracts/governance/NextyManager.sol]:
```suggestion
    ERC20 public ntfAddress;
```

File: contracts/governance/NextyManager.sol [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    //voting smart contract to lock(bann) or unlock(unbann) a sealer
+    address public voteAddress;
+
+    //get invester address of a coinbase
+    mapping(address => Record) public sealers;
+
+    //address[] public coinbases;
```
Comment (hadv) [inline contracts/governance/NextyManager.sol]: remove unused code please, dont comment out like this

File: contracts/governance/NextyManager.sol [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+                            //Sealer casted out of activation sealer set
+
+        WITHDRAWN,          //Sealer already withdrawn their deposit NTFs successfully. 
+                            //They can only make withdrawal after withdrawal period.
+
+        PENALIZED           //Sealer marked as penalized node (update by consensus or voting result via dapp) 
+                            //and cannot become active sealer and cannot withdraw balance neither.
+    }
+
+    struct Record {
```
Comment (hadv) [inline contracts/governance/NextyManager.sol]: rename the struct as `SealerRecord` for more clear

File: contracts/governance/NextyManager.sol [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        }
+    }
+
+    /**
+    * Transfer the NTF from token holder to registration contract. 
+    * Sealer might have to approve contract to transfer an amount of NTF before calling this function.
+    * @param tokens NTF Tokens to deposit
+    */
+    function deposit(uint256 tokens) public returns (bool) {
+        NTFToken(ntfAddress).transferFrom(msg.sender, address(this), tokens);
```
Comment (hadv) [inline contracts/governance/NextyManager.sol]: When declare `ntfAddress` as **ERC20** then you don't need to do type cast like this anymore

File: contracts/governance/NextyManager.sol [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        delete getSealer[_coinbase];
+        removeCoinbase(_coinbase);
+        emit Left(msg.sender, _coinbase);
+        return true;
+    }
+
+    /**
+    * To withdraw sealer’s NTF balance when they already exited and after withdrawal period.
+    */
+    function withdraw() public notBanned withdrawable returns (bool) {
```
Comment (hadv) [inline contracts/governance/NextyManager.sol]: 🆘 Must update state before calling transfer for security reason. ⚠️ 👨‍✈️

File: .eslintrc [hunk 8]
```diff
@@ -1,51 +1,35 @@
...
-      "assert": false,
-      "web3": false
-    },
-    "rules": {
-  
-      // Strict mode
-      "strict": [2, "global"],
-  
-      // Code style
-      "indent": [2, 2],
```
Comment (hadv) [inline .eslintrc]: put back the indent check to make code consistent; don't remove coding style rule

File: contracts/governance/SafeMath.sol [hunk 9]
```diff
@@ -0,0 +1,32 @@
+pragma solidity ^0.4.23;
```
Comment (hadv) [inline contracts/governance/SafeMath.sol]: if you don't have any additions, the import SafeMaths from openzeppline or zos-lib

**Comment by hadv** [comment]
⚠️Should adding at least the smoke tests for the nexty governance solidity smart contract to ensure the correctness when you refactoring the source code @bestboyvn87

File: README.md (hunk 1 above)
Comment (bestboyvn87) [reply README.md]: done

File: contracts/governance/NextyManager.sol (hunk 2 above)
Comment (bestboyvn87) [reply contracts/governance/NextyManager.sol]: done

File: contracts/governance/NextyManager.sol (hunk 4 above)
Comment (bestboyvn87) [reply contracts/governance/NextyManager.sol]: done

File: contracts/governance/NextyManager.sol (hunk 5 above)
Comment (bestboyvn87) [reply contracts/governance/NextyManager.sol]: done

File: contracts/governance/NextyManager.sol (hunk 7 above)
Comment (bestboyvn87) [reply contracts/governance/NextyManager.sol]: done

File: contracts/governance/SafeMath.sol (hunk 9 above)
Comment (bestboyvn87) [reply contracts/governance/SafeMath.sol]: done

File: .eslintrc (hunk 8 above)
Comment (bestboyvn87) [reply .eslintrc]: done

File: contracts/governance/NextyManager.sol [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+    //voting smart contract to lock(bann) or unlock(unbann) a sealer
+    address public voteAddress;
+
+    //get invester address of a coinbase
+    mapping(address => Record) public sealers;
+
+    //address[] public coinbases;
+    address[] public coinbases;
+    //check if address of a coinbase
+    mapping(address => bool) public isCoinbase;
```
Comment (hadv) [inline contracts/governance/NextyManager.sol]: dont use tobe as variable name, it make confusing with function name

File: contracts/governance/NextyManager.sol [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+    //get sealer's address of a coinbase address
+    mapping(address => address) public getSealer;
```
Comment (hadv) [inline contracts/governance/NextyManager.sol]: dont use verb as variable name, it make confusing with function name

File: contracts/governance/NextyManager.sol (hunk 10 above)
Comment (Zergity) [reply contracts/governance/NextyManager.sol]: done

File: contracts/governance/NextyManager.sol (hunk 11 above)
Comment (Zergity) [reply contracts/governance/NextyManager.sol]: done

File: contracts/governance/NextyGovernance.sol [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+    event Unbanned(address _sealer);
+
+    /**
+    * Check if address is a valid destination to transfer tokens to
+    * - must not be zero address
+    * - must not be the token address
+    * - must not be the sender's address
+    */
+    modifier validCoinbase(address _coinbase) {
+        require(cbSealer[_coinbase] != address("0x0"), "coinbase already used");
```
Comment (hadv) [inline contracts/governance/NextyGovernance.sol]: seems we have to use like this `address(0x0)`

```suggestion
        require(cbSealer[_coinbase] != address(0x0), "coinbase already used");
```

File: scripts/test.sh [hunk 13]
```diff
@@ -1,70 +1,70 @@
-#!/usr/bin/env bash
```
Comment (hadv) [inline scripts/test.sh]: Why do you comment all the test script? @bestboyvn87 ???

File: README.md (hunk 1 above)
Comment (hadv) [reply README.md]: because we dont use zos anymore, then now we need to update README. The previous version  of README now out-dated already
